packages/headless/src/features/commerce/context/cart/cart-slice.ts

```diff
@@ -0,0 +1,36 @@
+import {createReducer} from '@reduxjs/toolkit';
+import {CartState, getCartInitialState} from './cart-state';
+import {addItem, removeItem, setItems, updateItemQuantity} from './cart-actions';
+
+export const cartReducer = createReducer(
+  getCartInitialState(),
+
+  (builder) => {
+    builder
+      .addCase(setItems, (state, {payload}) => {
+        const {cart, cartItems} = payload.cart.reduce((acc, item) => {
+          return ({
+            cartItems: [...acc.cartItems, item.productId],
+            cart: {
+              ...acc.cart,
+              [item.productId]: item
+            }
+          }) as CartState;
+        }, getCartInitialState());
+
+        state.cartItems = cartItems;
+        state.cart = cart;
+      })
+      .addCase(addItem, (state, {payload}) => {
```

Comment:
I see these are now marked as outdated comments from my previous review, but they would still apply:

There need to be defensive code against productId being invalid/absent before we do any operation in the reducers.

Reply:
I've asked about possibly changing the methods we expose on the controller, to limit how defensive we need to be here: https://github.com/coveo/ui-kit/pull/3190#discussion_r1337658071